cli/cliutil/sink.go

```diff
@@ -0,0 +1,38 @@
+package cliutil
+
+import (
+	"io"
+	"sync"
+)
+
+type discardAfterClose struct {
+	sync.Mutex
+	wc     io.WriteCloser
+	closed bool
+}
+
+// DiscardAfterClose is an io.WriteCloser that discards writes after it is closed without errors.
+// It is useful as a target for a slog.Sink such that an underlying WriteCloser, like a file, can
+// be cleaned up without race conditions from still-active loggers.
+func DiscardAfterClose(wc io.WriteCloser) io.WriteCloser {
+	return &discardAfterClose{wc: wc}
+}
+
+func (d *discardAfterClose) Write(p []byte) (n int, err error) {
+	d.Lock()
+	defer d.Unlock()
+	if d.closed {
+		return len(p), nil
```

Comment:
Does slog print errors if we return io.EOF here? If yes and since this writer has no way to inform routines using it to stop, should we rename it something slog specific since we're essentially working around the fact that slog doesn't let you close it or close/detach sinks.

Also, if we need this at all, it means we're leaving goroutines behind when exciting, right? Often it's not a problem but it could mean we're not shutting down things in the proper order. This would essentially hide that.

Reply:
> Does slog print errors if we return io.EOF here?

the `Logger` does not, but some of the `Sinks` call `println` if they get an error writing, which I'd like to avoid.

> If yes and since this writer has no way to inform routines using it to stop, should we rename it something slog specific since we're essentially working around the fact that slog doesn't let you close it or close/detach sinks.

I did mention slog in the comment on the constructor, but I'm open to a name change if you have a suggestion.

> Also, if we need this at all, it means we're leaving goroutines behind when exciting, right? Often it's not a problem but it could mean we're not shutting down things in the proper order. This would essentially hide that.

Yes, many of our components just use a `Context` expiring to know when to shut down, and others that have an explicit `Close()` don't wait for all the goroutines to exit before returning.  These components are still "correct" in the sense that the goroutines do exit eventually---they're not leaks, just shut down is async with no way to wait for it.

We could mandate that everything has a `Close()` and it doesn't return until every goroutine is shut down, but that's pretty heavy handed and a lot of work.

If there are actual leaked goroutines, then we have `goleak` to detect those in tests. It's not worth trying to use the logger to police that.